crates/ty_python_semantic/resources/mdtest/typed_dict.md

```diff
@@ -1726,5 +1726,268 @@ reveal_type(actual_td)  # revealed: ActualTypedDict
+common need to have *compatible* types (in the fully-static case, equivalent types):
+
+```py
+from typing import Any, Generic, TypeVar
+
+class IntTD(TypedDict):
+    x: int
+
+class BoolTD(TypedDict):
+    x: bool
+
+# `bool` is assignable to `int`, but `int` is not assignable to `bool`. If `x` was `ReadOnly` (even,
+# as we'll see below, only on the `int` side), then these two TDs would not be disjoint, but in this
+# mutable case they are.
+
+static_assert(is_disjoint_from(IntTD, BoolTD))
+static_assert(is_disjoint_from(BoolTD, IntTD))
+
+# Gradual types: `int` is compatible with `bool | Any`, because that could materialize to
+# `bool | int`, which is just `int`. (And `int | Any` and `bool | Any` are compatible with each
+# other for the same reason.) However, `bool` is *not* compatible with `int | Any`, because there's
+# no materialization that's equivalent to `bool`.
+
+class IntOrAnyTD(TypedDict):
+    x: int | Any
+
+class BoolOrAnyTD(TypedDict):
+    x: bool | Any
+
+static_assert(not is_disjoint_from(IntTD, IntOrAnyTD))
+static_assert(not is_disjoint_from(IntOrAnyTD, IntTD))
+static_assert(not is_disjoint_from(IntTD, BoolOrAnyTD))
+static_assert(not is_disjoint_from(BoolOrAnyTD, IntTD))
+
+static_assert(not is_disjoint_from(IntOrAnyTD, BoolOrAnyTD))
+static_assert(not is_disjoint_from(BoolOrAnyTD, IntOrAnyTD))
+
+static_assert(is_disjoint_from(BoolTD, IntOrAnyTD))
+static_assert(is_disjoint_from(IntOrAnyTD, BoolTD))
+static_assert(not is_disjoint_from(BoolTD, BoolOrAnyTD))
+static_assert(not is_disjoint_from(BoolOrAnyTD, BoolTD))
+
+# `Any` is compatible with everything.
+
+class AnyTD(TypedDict):
+    x: Any
+
+static_assert(not is_disjoint_from(IntTD, AnyTD))
+static_assert(not is_disjoint_from(AnyTD, IntTD))
+static_assert(not is_disjoint_from(BoolTD, AnyTD))
+static_assert(not is_disjoint_from(AnyTD, BoolTD))
+static_assert(not is_disjoint_from(IntOrAnyTD, AnyTD))
+static_assert(not is_disjoint_from(AnyTD, IntOrAnyTD))
+static_assert(not is_disjoint_from(BoolOrAnyTD, AnyTD))
+static_assert(not is_disjoint_from(AnyTD, BoolOrAnyTD))
+static_assert(not is_disjoint_from(AnyTD, AnyTD))
+
+# This works with generic `TypedDict`s too.
+
+class TwoIntsTD(TypedDict):
+    x: int
+    y: int
+
+class TwoBoolsTD(TypedDict):
+    x: bool
+    y: bool
+
+class IntBoolTD(TypedDict):
+    x: int
+    y: bool
+
+T = TypeVar("T")
+
+class TwoGenericTD(TypedDict, Generic[T]):
+    x: T
+    y: T
+
+static_assert(not is_disjoint_from(TwoGenericTD[Any], TwoIntsTD))
+static_assert(not is_disjoint_from(TwoGenericTD[int], TwoIntsTD))
+static_assert(is_disjoint_from(TwoGenericTD[bool], TwoIntsTD))
+static_assert(not is_disjoint_from(TwoGenericTD[Any], TwoBoolsTD))
+static_assert(is_disjoint_from(TwoGenericTD[int], TwoBoolsTD))
+static_assert(not is_disjoint_from(TwoGenericTD[bool], TwoBoolsTD))
+# TODO: T can't be compatible with both `int` and `bool` at the same time, so these types should be
+# disjoint, regardless of the materialization of `T`.
```

Comment:
Is this sort of "one dynamic type winds up in multiple fields" situation ever something we reason about? Is there a name for this?